Improvements to processing of crop calendar files

samsrabin opened the pull request:
### Description of changes

In `python/ctsm/crop_calendars/`, `process_ggcmi_shdates.py` and `regrid_ggcmi_shdates.py` are used to convert the raw GGCMI crop calendar files into CTSM-compatible versions. This update fixes some bugs, removes dependencies on the `nco` utilities, and enables the use of surface datasets as template files (in addition to the existing ability to use CTSM output files).

### Specific notes

Contributors other than yourself, if any: None

CTSM Issues Fixed (include github issue #): None

Are answers expected to change (and if so in what way)? No

Any User Interface Changes (namelist or namelist defaults changes)? No

Testing performed, if any: I have used the scripts successfully.

ekluzek commented on python/ctsm/crop_calendars/regrid_ggcmi_shdates.py:
```diff
@@ -32,47 +33,106 @@ def define_arguments(parser):
-    for cmd in ["ncks", "ncrename", "ncpdq", "cdo"]:
+    # Ensure we can call necessary shell script(s)
+    for cmd in ["cdo"]:
         run_and_check(f"{cmd} --help")
```
This is an interesting and nice feature to check binaries that are needed. I like this.

ekluzek commented on python/ctsm/crop_calendars/regrid_ggcmi_shdates.py:
```diff
@@ -32,47 +33,106 @@ def define_arguments(parser):
 
 
+# lat or lon
+def import_coord_1d(ds, coordName):
```
I like putting this into a standalone function. This means you can easily unit test it, and
potentially use it elsewhere.

Are there other places that could use this function? If so it possibly could be moved to
a more generic place. Or it might be fine to leave it here, and you could just use this
this function from this module.

ekluzek replied:
I would like to see some unit tests added for this. There are some things that could be easily unit tested here. It's good to add unit tests for error checks, as they lend themselves to unit tests, but not to system tests usually.

There is also the system test that's run and that will test a lot of the code for correctness. So the unit tests could supplement those tests for parts of the code that aren't tested in the rxcropmaturity.py system test.

samsrabin replied:
- Moved to python/ctsm/utils.py in 51cc972.
- Unit testing added in 96b88aa and 04d2d73.

ekluzek commented on python/ctsm/crop_calendars/regrid_ggcmi_shdates.py:
```diff
@@ -32,47 +33,106 @@ def define_arguments(parser):
+
+
+# LATIXY or LONGXY
+def import_coord_2d(ds, coordName, varName):
```
Same thoughts as above.

samsrabin replied:
- Moved to python/ctsm/utils.py in 51cc972.
- Unit testing added in 96b88aa.

ekluzek commented on python/ctsm/crop_calendars/regrid_ggcmi_shdates.py:
```diff
@@ -32,47 +33,106 @@ def define_arguments(parser):
+    thisDim = thisDim[0]
+    otherDim = [x for x in da.dims if coordName not in x]
+    if len(otherDim) != 1:
+        raise RuntimeError(
```
Something that could use a unit test for a error check.

samsrabin replied:
Done in 96b88aa.

ekluzek commented on python/ctsm/crop_calendars/regrid_ggcmi_shdates.py:
```diff
@@ -32,47 +33,106 @@ def define_arguments(parser):
+    da = ds[varName]
+    thisDim = [x for x in da.dims if coordName in x]
+    if len(thisDim) != 1:
+        raise RuntimeError(
```
Could have a unit error check test.

samsrabin replied:
Done in 96b88aa.

ekluzek reviewed (changes requested):
I like this work, the code changes look great to me. There are two things I wonder about. First if the functions being added might be useful in other contexts. And as such I wonder if they could be used from where they are, or if it would be good to move to a utility module.

Second, I think it would be worth adding some unit testing for this. This would be good to add more tests under the normal python testing. And would also be good to add tests for the error checking in the code. Error checks are important to have, but can easily be incorrect without testing to make sure they are right.

I think we should talk over this and decide if that makes sense. This could also be done later. But, would also be good to get in now with the code.

samsrabin commented:
@ekluzek This is ready for re-review. I moved the functions in question to python/ctsm/utils.py and added unit tests for them, and I also added a system test for `regrid_ggcmi_shdates`.

slevis-lmwg commented:
> @slevis-lmwg Do you have `python/ctsm/test/testinputs/cropcals/swh_rf_ggcmi_crop_calendar_phase3_v1.01.nc4` in your local copy?

I have it.

slevis-lmwg commented:
> @slevis-lmwg Could you paste the entire error message?

Here's everything, starting with the command that I'm executing:
```
(ctsm_pylib) slevis@derecho1:/glade/work/slevis/git/single_point/python> ./run_ctsm_py_tests -s
................
Inactive Modules:
  1) cray-mpich/8.1.25     2) hdf5/1.12.2     3) intel/2023.0.0     4) ncarcompilers/1.0.0     5) netcdf/4.9.2

Due to MODULEPATH changes, the following have been reloaded:
  1) conda/latest     2) craype/2.7.20

The following have been reloaded with a version change:
  1) ncarenv/23.06 => ncarenv/23.09     2) nco/5.1.4 => nco/5.1.6

The following modules were not unloaded:
  (Use "module --force purge" to unload all):

  1) cesmdev/1.0   2) ncarenv/23.09
Done converting /glade/derecho/scratch/slevis/tmp/tmpin1vokuj/scrip.nc
..E
Stdout:
Regridding GGCMI crop calendars to 5x5amazon:
.
======================================================================
ERROR: test_regrid_ggcmi_shdates (test.test_sys_regrid_ggcmi_shdates.TestRegridGgcmiShdates)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/glade/work/slevis/git/single_point/python/ctsm/test/test_sys_regrid_ggcmi_shdates.py", line 91, in test_regrid_ggcmi_shdates
    args.crop_list,
  File "/glade/work/slevis/git/single_point/python/ctsm/crop_calendars/regrid_ggcmi_shdates.py", line 64, in main
    run_and_check(f"{cmd} --help")
  File "/glade/work/slevis/git/single_point/python/ctsm/crop_calendars/regrid_ggcmi_shdates.py", line 29, in run_and_check
    abort(f"Trouble running `{result.args}` in shell:\n{result.stdout}\n{result.stderr}")
  File "/glade/work/slevis/git/single_point/python/ctsm/utils.py", line 27, in abort
    sys.exit("ERROR: {}".format(errmsg))
SystemExit: ERROR: Trouble running `cdo --help` in shell:

shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
/bin/sh: cdo: command not found


Stdout:
Regridding GGCMI crop calendars to 5x5amazon:

----------------------------------------------------------------------
Ran 20 tests in 63.274s

FAILED (errors=1)
```

samsrabin commented:
@slevis-lmwg I'm seeing `/bin/sh: cdo: command not found`; try `module load cdo` first.

slevis-lmwg commented:
> @slevis-lmwg I'm seeing `/bin/sh: cdo: command not found`; try `module load cdo` first.

Perfect, thank you. That took care of it. I'm adding instructions to that effect in the corresponding README.

ekluzek commented:
@samsrabin I'm not familiar with CDO, but it seems it might be similar to NCO? What are the things it brings that are helpful here? I little bit of documentation would be helpful here.

ekluzek commented on python/ctsm/crop_calendars/regrid_ggcmi_shdates.py:
```diff
@@ -118,16 +177,45 @@ def main(
+        "--crop-list",
+        help="List of GGCMI crops to process; e.g., '--crop-list mai_rf,mai_ir'. If not provided, will process all GGCMI crops.",
+        default=None,
+    )
```
Oh, addendum, I see the error check for crop_list above on now line 128. Still adding the valid_values to the add_argument is also good to add. It's really helpful to the user.

samsrabin replied:
Avoiding the specification of valid values allows the script to be as flexible as possible, minimizing the need for future updates. My thought was that the provided examples would be helpful enough. What do you think?

ekluzek commented on python/ctsm/test/test_unit_regrid_ggcmi_shdates.py:
```diff
@@ -0,0 +1,125 @@
+    def test_importcoord1d(self):
+        ds = xr.open_dataset(self._1d_lonlat_file)
+        lat, Nlat = import_coord_1d(ds, "lat")
+        print(lat)
```
Remove this print in the test

samsrabin replied:
Done with 27b2607.

ekluzek commented on python/ctsm/crop_calendars/regrid_ggcmi_shdates.py:
```diff
@@ -32,49 +42,95 @@ def define_arguments(parser):
+        if crop_list is not None and this_crop not in crop_list:
+            continue
+
+        print("    " + this_crop)
```
Sam with this print as above about logging.

samsrabin replied:
Done with f39f7b4.

ekluzek commented on python/ctsm/utils.py:
```diff
@@ -191,6 +192,35 @@ def write_output(file, file_in, file_out, file_type):
+    da = ds[varName]
+    thisDim = [x for x in da.dims if coordName in x]
+    if len(thisDim) != 1:
+        abort(f"Expected 1 dimension name containing {coordName}; found {len(thisDim)}: {thisDim}")
```
I made sure there was an error check in the tester for this and there is! So awesome.

ekluzek commented on python/ctsm/utils.py:
```diff
@@ -191,6 +192,35 @@ def write_output(file, file_in, file_out, file_type):
+    otherDim = [x for x in da.dims if coordName not in x]
+    if len(otherDim) != 1:
+        abort(
+            f"Expected 1 dimension name not containing {coordName}; found {len(otherDim)}: {otherDim}"
```
Same as above. Checking the error checking is important in order to get it right.

ekluzek commented:
@samsrabin and @slevis-lmwg I went over this and it turns out I do have some suggestions. So I did my review.

I suggest for moving forward the following:
- @samsrabin and I meet to discuss things that can be done immediately
- We file issues on things to look into for longer term
- @samsrabin does the simple things in this PR
- @slevis-lmwg brings in those updates when they are done

Since this isn't FORTRAN code changes, this can be checked on with python testing, and maybe rerunning a couple run_sys_tests specific tests, so not a big deal. If it was FORTRAN changes we'd need to pause the merge tag work, but I don't think that's needed here. The other option is to back this out of the merge tag, but I don't think we need that either.

If you have thoughts on my suggestions please let us know, otherwise I think we should plan on the above.

samsrabin commented:
@ekluzek With AGU next week there's nothing that can be done immediately, unfortunately—I'm working on that all day today. Let's just back this out of the merge tag @slevis-lmwg is working on.

ekluzek commented:
Ahh, AGU, yes you are right. OK, this can wait. Part of the idea of the merge-tags and dev branch is that we easily delay things when needed. So it's good to normalize that in our process. 

@slevis-lmwg let us know if you'd like some advice on how to back-out this commit from your merge

slevis-lmwg commented on python/ctsm/crop_calendars/regrid_ggcmi_shdates.py:
```diff
@@ -32,49 +42,95 @@ def define_arguments(parser):
-    for cmd in ["ncks", "ncrename", "ncpdq", "cdo"]:
+    # Ensure we can call necessary shell script(s)
+    for cmd in ["cdo"]:
         run_and_check(f"{cmd} --help")
```
Here and elsewhere, reminder to add
`module load cdo;` before running cdo
to prevent sys test error.

samsrabin replied:
Done with 164357e.

slevis-lmwg reviewed (changes requested):
@samsrabin 
I'm adding this TODO to your PR, so that it may not be forgotten.

slevis-lmwg commented:
This PR switched to "merged" automatically because I originally merged it to my #2253 branch with f47de1d (even though I later reverted that commit).

I don't know the best way to reopen this PR, so I will let you @samsrabin do what you need to do. Sorry for the confusion...

ekluzek commented:
Ahhh, interesting. This is something we need to figure out in order to do merge tags and the development branch idea. So it's just a part of our learning about how to do this.

What that says is if you bring in a PR, but then revert that PR, it'll count it as merged. Which makes sense from git's point of view. I hadn't really thought it would work that way, but now it makes sense. So dropping a PR from a merge or development branch means we need to do something different. We should look into other options. Dropping the last PR is simple, but dropping a PR in the middle is more involved, especially if you can't just revert. So hmmm...

samsrabin commented on python/ctsm/test/test_sys_regrid_ggcmi_shdates.py:
```diff
@@ -0,0 +1,128 @@
+        self._regridded_cropcals = os.path.join(self._tempdir, "regridded_cropcals")
+
+        # What extension do the raw crop calendar file(s) have?
+        self._extension = ".nc4"
```
It's not the version that matters here, it's just that this script needs _an_ extension to look for when generating the list of files to process.

samsrabin commented on python/ctsm/test/test_sys_regrid_ggcmi_shdates.py:
```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python3
+
+"""System tests for fsurdat_modifier
```
Done with 197e53a

samsrabin commented on python/ctsm/utils.py:
```diff
@@ -191,6 +192,35 @@ def write_output(file, file_in, file_out, file_type):
     file.close()
 
 
+# Import 1-d latitude or longitude variable from a CESM history file (e.g., name lat or lon) and return it as a DataArray that can be used for writing CESM input files
```
Done with 197e53a.

samsrabin commented on python/ctsm/utils.py:
```diff
@@ -191,6 +192,35 @@ def write_output(file, file_in, file_out, file_type):
+    return da, len(da)
+
+
+# Import 1-d latitude or longitude variable from a CESM history file (e.g., name LATIXY or LONGXY) and return it as a 1-d DataArray that can be used as a coordinate for writing CESM input files
```
Done with 197e53a.

samsrabin commented on python/ctsm/crop_calendars/regrid_ggcmi_shdates.py:
```diff
@@ -91,13 +147,16 @@ def main(
-    ###############################
-    ### Process input arguments ###
-    ###############################
+# Process input arguments
```
Done with 0224c2a.

samsrabin commented on python/ctsm/test/test_unit_regrid_ggcmi_shdates.py:
```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python3
+
+"""
+Unit tests for regrid_ggcmi_shdates subroutines:
```
Done with 4bf2384.

samsrabin commented on python/ctsm/crop_calendars/regrid_ggcmi_shdates.py:
```diff
@@ -32,49 +42,95 @@ def define_arguments(parser):
+    extension,
+    crop_list,
 ):
+    print(f"Regridding GGCMI crop calendars to {regrid_resolution}:")
```
Done with f39f7b4.

samsrabin commented on python/ctsm/crop_calendars/regrid_ggcmi_shdates.py:
```diff
@@ -1,9 +1,21 @@
 import sys
+import xarray as xr
+import numpy as np
+
```
Done with 3ad5265df.
